src/kde/quick_index.cc

```diff
@@ -0,0 +1,813 @@
+//--------------------------------------------*-C++-*---------------------------------------------//
+/*!
+ * \file   kde/quick_index.cc
+ * \author Mathew Cleveland
+ * \brief  Explicitly defined quick_index functions.
+ * \note   Copyright (C) 2018-2020 Triad National Security, LLC.
+ *         All rights reserved. */
+//------------------------------------------------------------------------------------------------//
+
+#include "quick_index.hh"
+
+namespace rtt_kde {
+//------------------------------------------------------------------------------------------------//
+/*!
+ * \brief
+ * 
+ * quick_index constructor. This function builds up a global indexing table to
+ * quickly access data that is spatial located near each other. It breaks up
+ * the data into equally spaced bins in each dimension. For domain decomposed
+ * data it builds a one sided communication map to place local data that is
+ * need on other processors for ghost cells. The ghost cell extents is
+ * determined by the max_data_window spatial size such that any cell on the
+ * local domain will have access to all points that should fall into the
+ * spatial window centered on any given local point.
+ *
+ * \tparam dim integer specifying the data dimensionality 
+ * \param[in] data locations
+ * \param[in] domain_decomposed
+ */
+quick_index::quick_index(const size_t dim_, const std::vector<std::array<double, 3>> &locations_,
+                         const double max_window_size_, const size_t bins_per_dimension_,
+                         const bool domain_decomposed_)
+    : dim(dim_), domain_decomposed(domain_decomposed_), coarse_bin_resolution(bins_per_dimension_),
+      max_window_size(max_window_size_), locations(locations_), n_locations(locations_.size()) {
+
+  // Build local bounding box
+  bounding_box_min = {0, 0, 0};
+  bounding_box_max = {0, 0, 0};
+  // only set initial values for working dimensions
+  for (size_t d = 0; d < dim; d++) {
+    bounding_box_min[d] = 1e20;
+    bounding_box_max[d] = -1e20;
+  }
+
+  for (auto locItr = locations.begin(); locItr != locations.end(); locItr++) {
```

Comment:
A ranged based for loop should eliminate the need for the dereference of `locItr` to `loc`
```
for (const auto& loc : locations)
```